src/com/amazon/ionelement/api/Equivalence.kt

```diff
@@ -0,0 +1,127 @@
+ */
+public fun areElementsEqual(left: IonElement, right: IonElement): Boolean {
+    return left.asAnyElement().isEquivalentTo(right.asAnyElement())
+}
+
+/**
+ * Checks if two [StructField]s are equal.
+ *
+ * This function is normative for equality of [StructField]s.
+ * All [StructField] implementations must override [Any.equals] in a way that is equivalent to calling this function.
+ */
+public fun areFieldsEqual(left: StructField, right: StructField): Boolean = left.name == right.name && left.value == right.value
+
+/**
+ * Checks if this [AnyElement] is equal to some other object.
+ */
+internal fun AnyElement.isEquivalentTo(other: Any?): Boolean {
+    if (this === other) return true
+    if (other !is AnyElement) return false
+    return this.isEquivalentTo(other)
+}
+
+/**
+ * Internal only function that is equivalent to [areElementsEqual]
+ */
+private fun AnyElement.isEquivalentTo(other: AnyElement): Boolean {
+    if (this === other) return true
+    val thisType = this.type
+    if (thisType != other.type) return false
+    if (annotations != other.annotations) return false
+    // Metas intentionally not included here.
+
+    return if (isNull) {
+        // Already verified that they are the same type
+        other.isNull
+    } else if (other.isNull) {
+        false
+    } else
+    // Matching an enum rather than a type allows the Kotlin compiler
+    // to use a table switch instead of a chain of if/else comparisons.
+        when (thisType) {
+            BOOL -> booleanValue == other.booleanValue
+            INT -> when {
+                integerSize != other.integerSize -> false
+                integerSize == IntElementSize.LONG -> longValue == other.longValue
+                else -> bigIntegerValue == other.bigIntegerValue
+            }
+            // compareTo() distinguishes between 0.0 and -0.0 while `==` operator does not.
+            FLOAT -> doubleValue.compareTo(other.doubleValue) == 0
+            // `==` considers `0d0` and `-0d0` to be equivalent.  `Decimal.equals` does not.
+            DECIMAL -> Decimal.equals(decimalValue, other.decimalValue)
+            TIMESTAMP -> timestampValue == other.timestampValue
+            STRING -> stringValue == other.stringValue
+            SYMBOL -> symbolValue == other.symbolValue
+            BLOB -> blobValue == other.blobValue
+            CLOB -> clobValue == other.clobValue
+            LIST -> listValues == other.listValues
+            SEXP -> sexpValues == other.sexpValues
+            STRUCT -> {
+                val thisFields = this.structFields
+                val otherFields = other.structFields
+                when {
+                    thisFields === otherFields -> true
+                    thisFields.size != otherFields.size -> false
+                    // We've tried the inexpensive checks, now do a deep comparison
+                    else -> thisFields.groupingBy { it }.eachCount() == otherFields.groupingBy { it }.eachCount()
```

Comment:
is this comparing the occurrences of each field or am i missing something?

Reply:
It compares a `Map<StructField, Int>`. So for something like this Ion:
```
{ a: b, a: b, c: d }
```
It would create a map like this, which allows two structs to be compared even if their fields are in different orders, without having to impose a total ordering over all Ion values.
```
{
  a:b = 2, 
  c:d = 1,
}
```
It's not a one-way door. If it proves to be too costly of a comparison, we could (internally) create a total order over all `IonElement` so that we can compare two sorted lists of `StructField`s.

Reply:
Right... `it` is the key _and_ the value. I read it as being _just_ the key.